includes/class-event.php

```diff
@@ -0,0 +1,399 @@
+<?php
+
+namespace Automattic\WP\Cron_Control;
+
+use WP_Error;
+
+class Event {
+	private ?int $id;
+	private string $status;
+
+	// TODO: Maybe we don't need action_hashed going forward?
+	private string $action;
+	private string $action_hashed;
+
+	private array $args = [];
+	private string $instance;
+
+	// These are left empty for one-time events.
+	private ?string $schedule;
+	private ?int $interval;
+
+	// When the event will run next.
+	private int $timestamp;
+
+	// Keeps track of what properties are changed.
+	private array $changed = [];
+
+	/**
+	 * Get an existing event if an ID or row data is passed, otherwise the event is new and empty.
+	 * Note: Use Event::get() when possible instead of initializing an existing event here.
+	 *
+	 * @param int|object $event Event to initialize.
+	 */
+	public function __construct( $event = null ) {
+		if ( is_int( $event ) && $event > 0 ) {
+			$event_stub = new \stdClass();
+			$event_stub->ID = $event;
+			self::populate_data( $event_stub );
+		} elseif ( ! empty( $event->ID ) ) {
+			self::populate_data( $event );
+		}
+	}
+
+	/*
+	|--------------------------------------------------------------------------
+	| Getters
+	|--------------------------------------------------------------------------
+	*/
+
+	public function get_id(): ?int {
+		return isset( $this->id ) ? $this->id : null;
+	}
+
+	public function get_status(): ?string {
+		return isset( $this->status ) ? $this->status : null;
+	}
+
+	public function get_action(): ?string {
+		return isset( $this->action ) ? $this->action : null;
+	}
+
+	public function get_args(): array {
+		return $this->args;
+	}
+
+	public function get_instance(): string {
+		// Defaults to a hash of the empty args array.
+		return isset( $this->instance ) ? $this->instance : self::create_instance_hash( $this->args );
+	}
+
+	public function get_schedule(): ?string {
+		return isset( $this->schedule ) ? $this->schedule : null;
+	}
+
+	public function get_interval(): ?int {
+		return isset( $this->interval ) ? $this->interval : null;
+	}
+
+	public function get_timestamp(): ?int {
+		return isset( $this->timestamp ) ? $this->timestamp : null;
+	}
+
+	/*
+	|--------------------------------------------------------------------------
+	| Setters
+	|--------------------------------------------------------------------------
+	*/
+
+	public function set_status( string $status ): void {
+		$status = strtolower( $status );
+
+		if ( in_array( $status, Events_Store::ALLOWED_STATUSES, true ) ) {
+			$this->status = $status;
+		} else {
+			$this->status = Events_Store::STATUS_PENDING;
```

Comment:
Whatcha mean?

Reply:
I mean if the status is NOT in allowed statuses, we discard what the caller send in and set PENDING. Should we instead throw an error and say something like: "The status `burn_in_hell` that you provided is not an allowed status."

Reply:
I guess it's a matter of defaults. Right now if you don't set a status, a new event is created as `pending` automatically. So my thought process was to just default a failed `set_status()` attempt as pending, as it's the safe (and only) core WP status. 

The rest of the setters kind of follow a pattern where they protect the defaults, and just won't set if given missing or invalid data, helping ensure data integrity w/o the consumer needs to check the response of each set() (instead only need to confirm if the event was able to be saved with `->save()`, which will fail if it didn't get the required fields).

That said, the behavior in set_status() is a bit different since it makes a more active decision in the process. I'm thinking could just remove the active decision and just not update the status when given an invalid one. If it's a new event, `STATUS_PENDING` will be applied automatically. If it's an existing event, then whatever it was before the invalid set attempt will stay present. That does seem more sane.